### PR TITLE
v9: Throw error on duplicate routes

### PR DESCRIPTION
**Pull request by Zeegaan**
fixes https://github.com/umbraco/Umbraco-CMS/issues/11584
# Notes
- Created IConflictingRouteService interface
- Created ConflictingRouteService class

# How to test
- Create 2 controllers that derive from `UmbracoAuthorizedApiController`, remember to use 2 different namespaces
- Run umbraco and assert that you now get an error.
- Please also test this with making a controller with the same name as one of the umbraco API controllers, like TourController

### CODE REVIEW

File: src/Umbraco.Web.BackOffice/Services/ConflictingRouteService.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        /// Initializes a new instance of the <see cref="ConflictingRouteService"/> class.
+        /// </summary>
+        public ConflictingRouteService(TypeLoader typeLoader) => _typeLoader = typeLoader;
+
+        /// <inheritdoc/>
+        public bool HasConflictingRoutes()
+        {
+            var controllers = _typeLoader.GetTypes<UmbracoApiControllerBase>().ToList();
+            // var pluginControllers = _typeLoader.GetTypes<PluginController>().ToList();
+            // controllers.AddRange(pluginControllers);
```
Comment (elit0451) [inline src/Umbraco.Web.BackOffice/Services/ConflictingRouteService.cs:23]: We can remove those right?
```suggestion

```

**Review by elit0451 — COMMENTED** [review]
Changes look good 🙂 
I am wondering if there is a way to show which controller is with a duplicate name and that way help people to easily fix the exception 🤔

File: src/Umbraco.Web.BackOffice/Services/ConflictingRouteService.cs (hunk 1 above)
Comment (Zeegaan) [reply src/Umbraco.Web.BackOffice/Services/ConflictingRouteService.cs:23]: Yep my bad 🙈

**Comment by Zeegaan** [comment]
> 

We could return a string with the ControllerName instead of a bool, and leave the string empty if there's no errors, something like
`if (!string.IsNullOrEmpty(_conflictingRouteService.HasConflictingRoutes()))`

**Comment by Zeegaan** [comment]
> I am wondering if there is a way to show which controller is with a duplicate name and that way help people to easily fix the exception 🤔

We could return a string with the ControllerName instead of a bool, and leave the string empty if there's no errors, something like
`if (!string.IsNullOrEmpty(_conflictingRouteService.HasConflictingRoutes()))`

**Comment by elit0451** [comment]
Or keep the name as an `out` variable?

**Comment by Zeegaan** [comment]
ah yes that is a better solution, let me fix that real quick

**Comment by elit0451** [comment]
Thanks for the fix! Merging

**Comment by keithswright** [comment]
Hi Guys, 

This looks like it solves the issue in this bug alright: https://github.com/umbraco/Umbraco-CMS/issues/11584
But it has introduced another issue for my site (perhaps I have implemented something poorly, so I am happy to be advised about how I have done it).

In my site, I have 3 different consumers of the same kind of data, but with different security requirements The data is queried from an API. These consumers are:
 - the site frontend
 - the umbraco backoffice
 - other custom backend services that query data from the CMS

so lets say I have some data called "Books". My folder structure looks like this:


**Controllers/BooksController.cs** 
(derived from UmbracoApiController and using a RouteAttribute, handles requests to /api/books)
**Controllers/Backoffice/BooksController.cs** 
(derived from UmbracoAuthorizedApiController and handles requests to /Umbraco/backoffice/api/books)
**Controllers/Private/BooksController.cs**
 (derived from UmbracoApiController and using a RouteAttribute, handles requests to /privateapi/books)

So, although I have 3 controllers with the same name, I am not conflicting with Umbraco's built in controller names, nor do I have a route conflict. But I get this this error on startup:

**Conflicting routes, you cannot have multiple controllers with the same name**

So, I wonder, would you expect my scenario above to work? 
Or is what I have done considered a bad pattern and I should rename all my controllers? 
Or move them into areas?

One way I can think of now, to get around this issue, and keep serving the same data on the same urls, is to rename the above controllers to the following, and keep my RouteAttributes the same.
**Controllers/**Frontend**BooksController.cs**
**Controllers/Backoffice/BooksController.cs** 
**Controllers/Private/**Private**BooksController.cs**

Which to doesn't feel ideal to me

Many thanks

**Comment by bergmania** [comment]
Sounds like a bug.. We will need to test it out an repro, and ensure your usecase works :)

**Comment by bergmania** [comment]
@keithswright .. I have looked into this, without finding a good way to fix all cases. Especially when you combine RouteAttributes and the umbraco routing.

Another workaround, if you don't like you change your controller names, is to implement a `NoopConflictingRouteService`

```cs
    public class NoopConflictingRouteService : IConflictingRouteService
    {
        public bool HasConflictingRoutes(out string controllerName)
        {
            controllerName = null;
            return false;
        }
    }
```

And register this implementation instead of the original

```cs
    public class TestComposer : IComposer
    {
        public void Compose(IUmbracoBuilder builder) => builder.Services.AddUnique<IConflictingRouteService, NoopConflictingRouteService>();
    }
```

**Pull request merged**